assets/js/modules/analytics-4/components/settings/SettingsEnhancedMeasurementSwitch.js

```diff
@@ -0,0 +1,196 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
+import {
+	ENHANCED_MEASUREMENT_ENABLED,
+	ENHANCED_MEASUREMENT_FORM,
+	MODULES_ANALYTICS_4,
+	PROPERTY_CREATE,
+	WEBDATASTREAM_CREATE,
+} from '../../datastore/constants';
+import EnhancedMeasurementSwitch from '../common/EnhancedMeasurementSwitch';
+import {
+	isValidPropertyID,
+	isValidPropertySelection,
+	isValidWebDataStreamID,
+	isValidWebDataStreamSelection,
+} from '../../utils/validation';
+const { useSelect, useDispatch } = Data;
+
+export default function SettingsEnhancedMeasurementSwitch( {
+	hasAnalytics4Access,
+} ) {
+	const isEnhancedMeasurementEnabled = useSelect( ( select ) =>
+		select( CORE_FORMS ).getValue(
+			ENHANCED_MEASUREMENT_FORM,
+			ENHANCED_MEASUREMENT_ENABLED
+		)
+	);
+
+	const propertyID = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getPropertyID()
+	);
+
+	const webDataStreamID = useSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getWebDataStreamID()
+	);
+
+	const isLoadingProperties = useSelect( ( select ) => {
+		return select( MODULES_ANALYTICS_4 ).isLoadingProperties( {
+			hasModuleAccess: hasAnalytics4Access,
+		} );
+	} );
+
+	const isLoadingWebDataStreams = useSelect( ( select ) => {
+		return select( MODULES_ANALYTICS_4 ).isLoadingWebDataStreams( {
+			hasModuleAccess: hasAnalytics4Access,
+		} );
+	} );
+
+	const isEnhancedMeasurementStreamEnabled = useSelect( ( select ) => {
+		if (
+			! isValidPropertyID( propertyID ) ||
+			! isValidWebDataStreamID( webDataStreamID ) ||
+			isLoadingProperties ||
+			isLoadingWebDataStreams
+		) {
+			return null;
+		}
+
+		return select( MODULES_ANALYTICS_4 ).isEnhancedMeasurementStreamEnabled(
+			propertyID,
+			webDataStreamID
+		);
+	} );
+
+	const isLoading = useSelect( ( select ) => {
+		if (
+			! isValidPropertySelection( propertyID ) ||
+			! isValidWebDataStreamSelection( webDataStreamID ) ||
+			isLoadingProperties ||
+			isLoadingWebDataStreams
+		) {
+			return true;
+		}
+
+		if (
+			propertyID === PROPERTY_CREATE ||
+			webDataStreamID === WEBDATASTREAM_CREATE
+		) {
+			return false;
+		}
+
+		return ! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
+			'getEnhancedMeasurementSettings',
+			[ propertyID, webDataStreamID ]
+		);
+	} );
+
+	const { setValues } = useDispatch( CORE_FORMS );
+	const { setEnhancedMeasurementStreamEnabled } =
+		useDispatch( MODULES_ANALYTICS_4 );
+
+	// If `isEnhancedMeasurementEnabled` is already defined in the first render, and either `PROPERTY_CREATE` or
```

Comment:
This is required to support the cases where this component is rendered after the user has actioned the creation flow, i.e. "Set up a new property" or "Set up a new web data stream" is selected and they've pressed "Confirm Changes", the settings UI goes into the loading state, and then returns to the form view to show the "Additional permissions required" modal - _or_, when they continue to the OAuth flow and then show the form again upon return.